slatedb/src/cached_object_store/object_store.rs

```diff
@@ -92,15 +151,51 @@ impl CachedObjectStore {
         })
     }
 
-    // TODO: implement the put with cache here
-    #[allow(unused)]
     async fn cached_put_opts(
         &self,
         location: &Path,
         payload: object_store::PutPayload,
         opts: object_store::PutOptions,
     ) -> object_store::Result<PutResult> {
-        self.object_store.put_opts(location, payload, opts).await
+        // Only cache if the cache_puts option is enabled
+        if !self.cache_puts {
+            // If caching is disabled, just write to the upstream object store without cloning
+            return self.object_store.put_opts(location, payload, opts).await;
+        }
+
+        // First, write to the upstream object store (cloning payload is cheap since it's just a Arc internally)
+        let result = self
+            .object_store
+            .put_opts(location, payload.clone(), opts)
+            .await?;
+
+        // Then, save to local cache if admission policy allows it
+        // Get metadata and attributes using our cached head method (this also saves the head to cache)
+        match self.cached_head(location).await {
+            Ok(meta) => {
+                let entry = self
+                    .cache_storage
+                    .entry(&meta.location, self.part_size_bytes);
+                if self.admission_picker.pick(entry.as_ref()).admitted() {
+                    // Convert PutPayload to stream and save parts to cache.
+                    // Note: cached_head() already saved the head, so we only need to save parts
+                    let stream =
+                        stream::iter(payload.into_iter()).map(Ok::<Bytes, object_store::Error>);
+
+                    // Save parts only, ignoring errors (cache failures shouldn't fail the PUT operation)
+                    self.save_parts_stream(entry, stream, 0).await.ok();
+                }
+            }
+            Err(e) => {
+                // Log warning but don't fail the PUT operation
+                warn!(
+                    "Failed to save head to disk cache for {} with. Error {}",
```

Comment:
To keep in line with error syntax:

```suggestion
                    "failed to save head to disk cache [location={}, error={:?}]",
```